collector/generatorreceiver/internal/topology/topology.go

```diff
@@ -1,16 +1,82 @@
 package topology
 
+import "fmt"
 
-
-type Topology struct  {
-	Services []ServiceTier `json:"services" yaml:"services"`
+type Topology struct {
+	Services map[string]*ServiceTier `json:"services" yaml:"services"`
 }
 
 func (t *Topology) GetServiceTier(serviceName string) *ServiceTier {
-	for _, v := range t.Services {
-		if v.ServiceName == serviceName {
-			return &v
+	return t.Services[serviceName]
+}
+
+func (t *Topology) ValidateServiceGraph(rootRoutes []RootRoute) error {
+	for _, rr := range rootRoutes {
+		err := t.validateDownstreamCalls(rr.Service, rr.Route)
+		if err != nil {
+			return err
 		}
 	}
 	return nil
-}
\ No newline at end of file
+}
+
+func (t *Topology) validateDownstreamCalls(service string, route string) error {
+	seenCalls := map[string]bool{service + route: true}
+	orderedCalls := []string{service + route}
+	err := t.traverseServiceGraph(service, route, seenCalls, orderedCalls)
+	// TODO: optimize so that we don't check same service/route combo twice
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *Topology) traverseServiceGraph(service string, route string, seenCalls map[string]bool, orderedCalls []string) error {
```

Comment:
Totally optional, but this method could be simplified quite a bit by eliminating the copies of `seenCalls` and `orderedCalls`:
- Instead of creating and passing `currentSeenCalls`, you can add `service+route` to `seenCalls` before the for loop, pass `seenCalls` into the recursive call, and remove `service+route` from `seenCalls` after for loop finishes
- Similarly, instead of creating `currentOrderedCalls` you can pass in `append(orderedCalls, ds+dr)` to the recursive call.

Together this becomes:
```go
func (t *Topology) traverseServiceGraph(service string, route string, seenCalls map[string]bool, orderedCalls []string) error {
	downstreamCalls := t.GetServiceTier(service).GetRoute(route).DownstreamCalls
	// already validated existence of all services/routes, so ^ is safe
	seenCalls[service+route] = true
	for ds, dr := range downstreamCalls {
		if seenCalls[ds+dr] {
			return fmt.Errorf(printServiceCycle(orderedCalls, ds+dr))
		}

		err := t.traverseServiceGraph(ds, dr, seenCalls, append(orderedCalls, ds+dr))
		if err != nil {
			return err
		}
	}
	delete(seenCalls, service+route)
	return nil
}
```

Reply:
Ah this is much nicer, thank you

Reply:
(For future reference, this approach only works here because this method is synchronous; async recursive methods get gnarly.)